net/net.go

```diff
@@ -97,3 +103,63 @@ type ConnectionError struct {
 func NewConnectionError(status, message string, cause error) *ConnectionError {
 	return &ConnectionError{Status: status, Message: message, Cause: cause}
 }
+
+// ReadFromWithDst reads one packet from `conn` into `b` and returns the number
+// of bytes read, the source address, and the destination IP address.  It enables
+// recovery of the destination IP, which is otherwise lost for UDP connections
+// that are bound to `0.0.0.0` or `::`.
+func ReadFromWithDst(conn net.PacketConn, b []byte) (n int, src *net.UDPAddr, dst net.IP, err error) {
+	var tmpSrc net.Addr
+	if conn.LocalAddr().Network() == "udp4" {
```

Comment:
I'm not a fan of this type switch because you already have per type code when you create the UDP services and because the `PacketConn` here is an incomplete type spec: you only accept the udp4 and udp6 ones.

Instead, you could introduce the concept of a `BindingPacketConn` with an `Accept()` method that returns the `BoundPacketConn` that I mention below.

You can have separate factory functions for IP 4 and 6, so you won't need the type switch. The higher-level concept will make the code easier to read because it will allow us to understand the different concerns separately.

Reply:
OK, I've fully separated the IPv4 and IPv6 factories.